keras_nlp/models/deberta/disentangled_self_attention.py

```diff
@@ -0,0 +1,376 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Disentangled self-attention layer."""
+
+import math
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.utils.keras_utils import clone_initializer
+
+
+class DisentangledSelfAttention(keras.layers.Layer):
+    """DisentangledSelfAttention layer.
+
+    This is an implementation of disentangled self-attention as described in the
+    paper ["DeBERTaV3: Improving DeBERTa using ELECTRA-Style Pre-Training with Gradient-Disentangled Embedding Sharing"](https://arxiv.org/abs/2111.09543).
+    Effectively, this layer implements Multi-Head Self Attention with relative
+    attention, i.e., to get the final attention score, we compute the
+    content-to-position and position-to-content attention scores, and add these
+    scores to the vanilla multi-head self-attention scores.
+
+    Args:
+        num_heads: int. Number of attention heads.
+        hidden_dim: int. Hidden dimension of the input, i.e., `hidden_states`.
+        max_position_embeddings: int, defaults to 512. The maximum input
+            sequence length.
+        bucket_size: int, defaults to 256. The size of the relative position
+            buckets. Generally equal to `max_sequence_length // 2`.
+        dropout: float, defaults to 0.1. Dropout probability.
+        kernel_initializer: string or `keras.initializers` initializer,
+            defaults to "glorot_uniform". The kernel initializer for
+            the dense layers.
+        bias_initializer: string or `keras.initializers` initializer,
+            defaults to "zeros". The bias initializer for the dense layers.
+    """
+
+    def __init__(
+        self,
+        num_heads,
+        hidden_dim,
+        max_position_embeddings=512,
+        bucket_size=256,
+        dropout=0.1,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        name=None,
```

Comment:
We are not totally consistent, but most layers we have do not include this, maybe leave it out?